pkg/cloud/aws/aws.go

```diff
@@ -0,0 +1,430 @@
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String(fmt.Sprintf("tag:%s", common.InstanceTag)),
+					Values: []*string{aws.String("true")},
+				},
+			},
+			NextToken: nextToken,
+		}
+
+		instances, err := a.client.DescribeInstances(input)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range instances.Reservations {
+			for _, i := range r.Instances {
+				if *i.State.Name != ec2.InstanceStateNameTerminated {
+					arr := automation.RessourceResults{
+						ID:     *i.InstanceId,
+						Region: a.region,
+					}
+
+					if i.PublicIpAddress != nil {
+						arr.PublicIP = *i.PublicIpAddress
+					}
+
+					var tags []string
+					var instanceName string
+					for _, v := range i.Tags {
+						tags = append(tags, fmt.Sprintf("%s=%s", *v.Key, *v.Value))
+
+						if *v.Key == "Name" {
+							instanceName = *v.Value
+						}
+					}
+
+					arr.Tags = strings.Join(tags, ",")
+					arr.Name = instanceName
+
+					result = append(result, arr)
+				}
+			}
+		}
+
+		nextToken = instances.NextToken
+		if nextToken == nil {
+			break
+		}
+	}
+
+	return result, nil
+}
+
+func (a *Aws) CreateServer(args automation.ServerArgs) (*automation.RessourceResults, error) {
+	image, err := a.lookupAMI("ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210621")
+	if err != nil {
+		return nil, err
+	}
+
+	pubKeyFile, err := ioutil.ReadFile(fmt.Sprintf("%s.pub", args.MinecraftResource.GetSSH()))
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := a.client.ImportKeyPair(&ec2.ImportKeyPairInput{
+		KeyName:           aws.String(fmt.Sprintf("%s-ssh", args.MinecraftResource.GetName())),
+		PublicKeyMaterial: pubKeyFile,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	instanceInput := &ec2.RunInstancesInput{
+		ImageId:      image,
+		KeyName:      key.KeyName,
+		InstanceType: aws.String(args.MinecraftResource.GetSize()),
+		MinCount:     aws.Int64(1),
+		MaxCount:     aws.Int64(1),
+		TagSpecifications: []*ec2.TagSpecification{
+			{
+				ResourceType: aws.String("instance"),
+				Tags: []*ec2.Tag{
+					{
+						Key:   aws.String("Name"),
+						Value: aws.String(args.MinecraftResource.GetName()),
+					},
+					{
+						Key:   aws.String(common.InstanceTag),
+						Value: aws.String("true"),
+					},
+				},
+			},
+		},
+	}
+
+	if args.MinecraftResource.GetVolumeSize() > 0 {
+		instanceInput.BlockDeviceMappings = []*ec2.BlockDeviceMapping{
+			{
+				DeviceName: aws.String("/dev/sda1"),
+				Ebs: &ec2.EbsBlockDevice{
+					VolumeSize: aws.Int64(int64(args.MinecraftResource.Spec.Server.VolumeSize)),
+				},
+			},
+		}
+	}
+
+	userData, err := a.tmpl.GetTemplate(args.MinecraftResource, "", minctlTemplate.GetTemplateCloudConfigName(args.MinecraftResource.IsProxyServer()))
+	if err != nil {
+		return nil, err
+	}
+	instanceInput.UserData = aws.String(base64.StdEncoding.EncodeToString([]byte(userData)))
+
+	groupID, _, err := a.createEC2SecurityGroup("", args.MinecraftResource.GetPort())
+	if err != nil {
+		return nil, err
+	}
```

Comment:
```go
	var secGroups []*string
	var groupID *string
	if args.MinecraftResource.GetEdition() == "bedrock" || args.MinecraftResource.GetEdition() == "nukkit" || args.MinecraftResource.GetEdition() == "powernukkit" {
		groupID, _, err = a.createEC2SecurityGroup("", "udp", args.MinecraftResource.GetPort())
		if err != nil {
			return nil, err
		}
	} else {
		groupID, _, err = a.createEC2SecurityGroup("", "tcp", args.MinecraftResource.GetPort())
		if err != nil {
			return nil, err
		}
		if args.MinecraftResource.HasRCON() {
			rconID, _, err := a.createEC2SecurityGroup("", "tcp", args.MinecraftResource.GetRCONPort())
			if err != nil {
				return nil, err
			}
			secGroups = append(secGroups, rconID)
		}
	}
	secGroups = append(secGroups, groupID)
	if args.MinecraftResource.HasMonitoring() {
		promGroupID, _, err := a.createEC2SecurityGroup("", "tcp", 9090)
		if err != nil {
			return nil, err
		}
		secGroups = append(secGroups, promGroupID)
	}
	sshGroupID, _, err := a.createEC2SecurityGroup("", "tcp", 22)
	if err != nil {
		return nil, err
	}
	secGroups = append(secGroups, sshGroupID)

	var networkSpec = ec2.InstanceNetworkInterfaceSpecification{
		DeviceIndex:              aws.Int64(int64(0)),
		AssociatePublicIpAddress: aws.Bool(true),
		DeleteOnTermination:      aws.Bool(true),
		Groups:                   secGroups,
	}
```

Reply:
made the suggested changes
thanks again for the review!